File: apps/rust/src/chapter15/service.rs
//! アプリケーション層。置き場からモデルを読み込んで予測する。HTTP を知らない。

use super::domain::{ModelStore, Movie, Passenger, Result};
use super::store::{SALES_MODEL, SURVIVAL_MODEL};

/// モデルの名前と、読み込めるかどうか。
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ModelHealth {
    pub name: &'static str,
    pub ready: bool,
}

/// 予測サービス。置き場は約束（trait）で受け取るので、テストではスタブに差し替えられる。
pub struct PredictionService {
    store: Box<dyn ModelStore + Send + Sync>,
}

impl PredictionService {
    /// 置き場を差し込んだサービスを作る。
    pub fn new(store: Box<dyn ModelStore + Send + Sync>) -> Self {
        PredictionService { store }
    }

    /// 映画の興行収入を予測する。
    pub fn predict_sales(&self, movie: Movie) -> Result<f64> {
        self.store.load_sales_model()?.predict_sales(movie)
    }

    /// 乗客が生存するかを予測する。
    pub fn predict_survival(&self, passenger: &Passenger) -> Result<bool> {
        self.store.load_survival_model()?.survives(passenger)
    }

    /// モデルごとに読み込めるかどうかを返す。
    pub fn health(&self) -> Vec<ModelHealth> {
        vec![
            ModelHealth {
                name: SALES_MODEL,
                ready: self.store.load_sales_model().is_ok(),
            },
            ModelHealth {
                name: SURVIVAL_MODEL,
                ready: self.store.load_survival_model().is_ok(),
            },
        ]
    }
}

File: apps/php/src/Chapter15/Service.php
<?php

declare(strict_types=1);

namespace GettingStartedMl\Chapter15;

/**
 * 置き場からモデルを読み込んで予測するアプリケーション層。HTTP を知らない。
 *
 * 置き場は `ModelStore` を実装していれば何でもよいので、テストでは偽物を渡せる。
 * 自分では状態を持たない。
 */
final readonly class Service
{
    public function __construct(private ModelStore $store)
    {
    }

    /** 映画の特徴量から興行収入を予測する。 */
    public function predictSales(Movie $movie): float
    {
        return $this->store->loadSalesModel()->predict($movie);
    }

    /** 乗客の特徴量から生存するかどうかを予測する。 */
    public function predictSurvival(Passenger $passenger): bool
    {
        return $this->store->loadSurvivalModel()->predict($passenger);
    }

    /**
     * モデルごとに、読み込めるかどうかを返す。並びは Domain::MODEL_NAMES に従う。
     *
     * @return array<string, bool>
     */
    public function health(): array
    {
        return [
            Domain::SALES_MODEL => $this->ready($this->store->loadSalesModel(...)),
            Domain::SURVIVAL_MODEL => $this->ready($this->store->loadSurvivalModel(...)),
        ];
    }

    /**
     * モデルを読み込めるかどうか。読み込めない理由がほかにあれば、そのまま投げる。
     *
     * @param callable(): object $load
     */
    private function ready(callable $load): bool
    {
        try {
            $load();

            return true;
        } catch (ModelNotFoundException) {
            return false;
        }
    }
}
